### PR TITLE
CORDA-3432 Feature/checkpoint table improvements

### PR DESCRIPTION
**Pull request by lankydan**

### CODE REVIEW

**Comment by lankydan** [comment]
run tests: slow

**Comment by lankydan** [comment]
Run tests: slow

File: core/src/main/kotlin/net/corda/core/internal/FlowIORequest.kt [hunk 1]
```diff
@@ -45,6 +45,7 @@ sealed class FlowIORequest<out R : Any> {
      * @property shouldRetrySend specifies whether the send should be retried.
      * @return a map from session to received message.
      */
+    //net.corda.core.internal.FlowIORequest.SendAndReceive
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/FlowIORequest.kt:48]: What is this line?  I assume we should remove.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt [hunk 2]
```diff
@@ -48,6 +49,12 @@ open class SharedNodeCmdLineOptions {
     )
     var devMode: Boolean? = null
 
+    @Option(
+            names = ["--pause-all-flows"],
+            description = ["Do not run any flows on startup. Sets all flows to paused, which can be unpaused via RPC."]
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:54]: Double-checking.  Is this "unpause via RPC" true in OS ?

File: core/src/main/kotlin/net/corda/core/internal/messaging/InternalCordaRPCOps.kt [hunk 3]
```diff
@@ -13,4 +14,11 @@ interface InternalCordaRPCOps : CordaRPCOps {
 
     /** Get all attachment trust information */
     val attachmentTrustInfos: List<AttachmentTrustInfo>
+
+    /**
+     * Resume a paused flow.
+     *
+     * @return whether the flow was successfully resumed.
+     */
+    fun unPauseFlow(id: StateMachineRunId): Boolean
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/messaging/InternalCordaRPCOps.kt:23]: Does this mean we cannot unpause via RPC in OS?  What about via shell ?

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 4]
```diff
@@ -1,97 +1,625 @@
...
 
+    // DBFlowResult and DBFlowException to be integrated with rest of schema
+    @Suppress("MagicNumber")
     override fun removeCheckpoint(id: StateMachineRunId): Boolean {
+        var deletedRows = 0
+        val flowId = id.uuid.toString()
+        deletedRows += deleteRow(DBFlowMetadata::class.java, DBFlowMetadata::flowId.name, flowId)
+        deletedRows += deleteRow(DBFlowCheckpointBlob::class.java, DBFlowCheckpointBlob::flowId.name, flowId)
+        deletedRows += deleteRow(DBFlowCheckpoint::class.java, DBFlowCheckpoint::flowId.name, flowId)
+//        resultId?.let { deletedRows += deleteRow(DBFlowResult::class.java, DBFlowResult::flow_id.name, it.toString()) }
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:373]: Commented out lines?  Looks like maybe from the "future".  Can we add a comment saying what these lines are.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt [hunk 5]
```diff
@@ -1,97 +1,625 @@
...
+     * A [DBFlowException] is created if [DBFlowCheckpoint.exceptionDetails] does not exist and the [Checkpoint] has an error attached to it.
+     * The existing [DBFlowException] is updated if [DBFlowCheckpoint.exceptionDetails] exists and the [Checkpoint] has an error.
+     * The existing [DBFlowException] is deleted if [DBFlowCheckpoint.exceptionDetails] exists and the [Checkpoint] has no error.
+     * Nothing happens if both [DBFlowCheckpoint] and [Checkpoint] are related to no errors.
+     */
+    // DBFlowException to be integrated with rest of schema
+    // Add a flag notifying if an exception is already saved in the database for below logic (are we going to do this after all?)
+    private fun updateDBFlowException(flowId: String, checkpoint: Checkpoint, now: Instant): DBFlowException? {
+        val exceptionDetails = (checkpoint.errorState as? ErrorState.Errored)?.let { createDBFlowException(flowId, it, now) }
+//        if (checkpoint.dbExoSkeleton.dbFlowExceptionId != null) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:508]: Likewise, if these lines are likely to be re-instated, add a comment explaining that.

File: node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt [hunk 6]
```diff
@@ -61,6 +54,7 @@ import java.lang.Integer.min
 import java.security.SecureRandom
 import java.time.Duration
 import java.util.HashSet
+import java.util.*
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:57]: Surprise this passes Detekt.

**Review by rick-r3 — COMMENTED** [review]
Couple of small cosmetic things.

Main sticking point is can OS un-pause a flow?  We gave them batch pausing at start up.  The command line option says to use RPC to un-pause.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt (hunk 2 above)
Comment (lankydan) [reply node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:54]: Will is working on the retry/unpausing in https://github.com/corda/corda/pull/6247#pullrequestreview-419888017

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 4 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:373]: Indeed, these are to be uncommented once we integrate `DBFlowResult `and `DBFlowException` with rest of the entities. There is currently the following comment on top of method:

```
// DBFlowResult and DBFlowException to be integrated with rest of schema
```
If not good enough we can change it/ move closer to the commented code.

File: node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt (hunk 5 above)
Comment (kyriathar) [reply node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:508]: Same as above, comment on top of method:

```
// DBFlowException to be integrated with rest of schema
// Add a flag notifying if an exception is already saved in the database for below logic (are we going to do this after all?)
```

can change if not good enough.

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**